homeassistant/components/vodafonegigabox_router/device_tracker.py

```diff
@@ -0,0 +1,184 @@
+"""Support for Vodafone Gigabox routers."""
+import logging
+import re
+import hashlib
+import hmac
+import time
+from random import random
+from collections import namedtuple
+
+import requests
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.device_tracker import (
+    DOMAIN,
+    PLATFORM_SCHEMA,
+    DeviceScanner,
+)
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_USERNAME): cv.string,
+    }
+)
+
+
+def get_scanner(hass, config):
+    """Validate the configuration and return a Vodafone Gigabox scanner."""
+    scanner = VodafoneDeviceScanner(config[DOMAIN])
+
+    return scanner
+
+
+Device = namedtuple("Device", ["name", "ip", "mac"])
+
+
+class VodafoneDeviceScanner(DeviceScanner):
+    """This class queries a Vodafone Gigabox router."""
+
+    CSRF_REGEX = re.compile(r"csrf_token = '(.*?)'")
+    VODAFONE_API_PASS = "$1$SERCOMM$"
+
+    def __init__(self, config):
+        """Initialize the scanner."""
+        self.host = config[CONF_HOST]
+        self.username = config[CONF_USERNAME]
+        self.password = config[CONF_PASSWORD]
+        self.csrf_token = None
+        self.login_uid = None
+        self.last_results = []
+
+    def scan_devices(self):
+        """Scan for new devices and return a list with found device IDs."""
+        self._get_devices()
+        return [client.mac for client in self.last_results]
+
+    def get_device_name(self, device):
+        """Return the name of the given device or None if we don't know."""
+        if not self.last_results:
+            return None
+        for client in self.last_results:
+            if client.mac == device:
+                return client.name.replace(" (2.4GHz)", "").replace(" (5GHz)", "")
+        return None
+
+    def get_extra_attributes(self, device):
+        """Return extra attributes for the given device or None if we don't know."""
+        if not self.last_results:
+            return None
+        for client in self.last_results:
+            if client.mac == device:
+                return {"ip": client.ip}
+        return None
+
+    def _get_csrf(self):
+        """Get CSRF token."""
+        csrf = requests.get(
```

Comment:
This is protocol specific information that should be extracted to a standalone PyPI lib.

However, this also seems to be based on webscraping. We no longer allow integrations based on webscraping as per https://github.com/home-assistant/architecture/blob/master/adr/0004-webscraping.md